ports/esp32/machine_pwm.c

```diff
@@ -208,7 +211,42 @@ static void configure_channel(machine_pwm_obj_t *self) {
     }
 }
 
+// Temporary workaround for ledc_find_suitable_duty_resolution function only being added in IDF V5.2
+#if ESP_IDF_VERSION < ESP_IDF_VERSION_VAL(5, 2, 0)
+static uint32_t ledc_find_suitable_duty_resolution(uint32_t src_clk_freq, uint32_t timer_freq) {
+    // This implementation is based on the one used in Micropython v1.23
+
+    // Find the highest bit resolution for the requested frequency
+    unsigned int freq = src_clk_freq;
+
+    int divider = (freq + timer_freq / 2) / timer_freq; // rounded
+    if (divider == 0) {
+        divider = 1;
+    }
+    float f = (float)freq / divider; // actual frequency
+    if (f <= 1.0) {
+        f = 1.0;
+    }
+    freq = (unsigned int)roundf((float)freq / f);
+
+    unsigned int res = 0;
+    for (; freq > 1; freq >>= 1) {
+        ++res;
+    }
+    if (res == 0) {
+        res = 1;
+    } else if (res > HIGHEST_PWM_RES) {
+        // Limit resolution to HIGHEST_PWM_RES to match units of our duty
+        res = HIGHEST_PWM_RES;
+    }
+
+    return res;
+}
+#endif
+
+
 static void set_freq(machine_pwm_obj_t *self, unsigned int freq, ledc_timer_config_t *timer) {
+    esp_err_t err;
```

Comment:
Any reason this was moved here, instead of just keeping it as it was, defined at the place of use below?

Reply:
Yes, the first place of use is now inside an #ifdef block.

Reply:
Oh, I see, I missed that it's also used for `ledc_timer_config()`.

I would have put this within the `if (freq != timer->freq_hz)` block to keep its scope limited, but here is also fine considering it's a generic error return variable.